nvme.c

```diff
@@ -221,67 +217,53 @@ void *nvme_realloc(void *p, size_t len)
 	return result;
 }
 
-static void *__nvme_alloc_huge(size_t len, bool *huge)
-{
-	void *p;
-
-	if (!posix_memalign(&p, getpagesize(), len)) {
-		*huge = false;
-		memset(p, 0, len);
-		return p;
-	}
-	return NULL;
-}
-
 #define HUGE_MIN 0x80000
 
-#ifdef CONFIG_LIBHUGETLBFS
-void nvme_free_huge(void *p, bool huge)
+void nvme_free_huge(void *p, size_t len)
 {
-	if (huge) {
-		if (p)
-			free_hugepage_region(p);
-	} else {
+	size_t _len = ROUND_UP(len, 0x1000);
+
+	if (_len < HUGE_MIN) {
 		free(p);
+		return;
 	}
+
+	munmap(p, _len);
 }
 
-void *nvme_alloc_huge(size_t len, bool *huge)
+void *nvme_alloc_huge(size_t len)
 {
+	size_t _len = ROUND_UP(len, 0x1000);
 	void *p;
 
-	if (len < HUGE_MIN)
-		return __nvme_alloc_huge(len, huge);
+	if (_len < HUGE_MIN) {
+		if (!posix_memalign(&p, getpagesize(), len)) {
+			memset(p, 0, len);
+			return p;
+		}
+	}
 
-	p = get_hugepage_region(len, GHR_DEFAULT);
-	if (!p)
-		return __nvme_alloc_huge(len, huge);
+	p = mmap(NULL, _len, PROT_READ | PROT_WRITE,
+		 MAP_ANONYMOUS | MAP_PRIVATE, -1, 0);
+	if (p < 0)
+		return NULL;
 
-	*huge = true;
-	return p;
-}
-#else
-void nvme_free_huge(void *p, bool huge)
-{
-	free(p);
-}
+	if (madvise(p, _len, MADV_HUGEPAGE) < 0) {
```

Comment:
Is it possible to just include the `MAP_HUGETLB` flag in the `mmap()` call instead of issuing another syscall?

Reply:
Yes, that doesn't make really sense.